library/src/main/kotlin/com/connectrpc/impl/ClientOnlyStream.kt

```diff
@@ -31,59 +29,16 @@ internal class ClientOnlyStream<Input, Output>(
         return messageStream.send(input)
     }
 
-    override suspend fun receiveAndClose(): ResponseMessage<Output> {
-        val resultChannel = messageStream.resultChannel()
+    override suspend fun receiveAndClose(): Output {
+        val resultChannel = messageStream.responseChannel()
         try {
             messageStream.sendClose()
-            // TODO: Improve this API for consumers.
-            // We should aim to provide ease of use for callers so they don't need to individually examine each result
-            // in the channel (headers, 1* messages, completion) and have to resort to fold()/maybeFold() to interpret
-            // the overall results.
-            // Additionally, ResponseMessage.Success and ResponseMessage.Failure shouldn't be necessary for client use.
-            // We should throw ConnectException for failure and only have users have to deal with success messages.
-            var headers: Headers = emptyMap()
-            var message: Output? = null
-            var trailers: Headers = emptyMap()
-            var code: Code? = null
-            var error: ConnectException? = null
-            for (result in resultChannel) {
-                result.maybeFold(
-                    onHeaders = {
-                        headers = it.headers
-                    },
-                    onMessage = {
-                        message = it.message
-                    },
-                    onCompletion = {
-                        code = it.code
-                        trailers = it.trailers
-                        val resultErr = it.cause
-                        if (resultErr != null) {
-                            error = if (resultErr is ConnectException) {
-                                resultErr
-                            } else {
-                                ConnectException(code ?: Code.UNKNOWN, message = error?.message, exception = error, metadata = trailers)
-                            }
-                        }
-                    },
-                )
+            val message = resultChannel.receive()
+            val additionalMessage = resultChannel.receiveCatching()
+            if (additionalMessage.isSuccess) {
+                throw ConnectException(code = Code.UNKNOWN, message = "unary stream has multiple messages")
```

Comment:
It seems strange to that we'd suggest the reason is unknown (per the code) when we know what's going on here.

Reply:
This matches the behavior from connect-go: https://github.com/connectrpc/connect-go/blob/96effedc8ac84da9f49392e5f7bdfee2e648247b/connect.go#L367C1-L368

I think in this case, someone has broken the contract of what a client streaming call should adhere to, so we return UNKNOWN. I don't think any other codes make sense for us to return here, as they're typically used by server implementations.